pkg/restore/log_client.go

```diff
@@ -0,0 +1,902 @@
+	})
+	return l.writeAndIngestPairs(ctx, kvs)
+}
+
+func (l *LogClient) applyKVChanges(ctx context.Context, tableID int64) error {
+	log.Debug("apply kv changes to tikv",
+		zap.Any("table", tableID),
+	)
+	dataKVs := cdclog.MakeRowsFromKvPairs(nil)
+	indexKVs := cdclog.MakeRowsFromKvPairs(nil)
+
+	tableBuffer := l.tableBuffers[tableID]
+	if tableBuffer.IsEmpty() {
+		log.Warn("no kv changes to apply")
+		return nil
+	}
+
+	var dataChecksum, indexChecksum cdclog.KVChecksum
+	for _, p := range tableBuffer.KvPairs {
+		p.ClassifyAndAppend(&dataKVs, &dataChecksum, &indexKVs, &indexChecksum)
+	}
+
+	err := l.writeRows(ctx, dataKVs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	dataKVs = dataKVs.Clear()
+
+	err = l.writeRows(ctx, indexKVs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	indexKVs = indexKVs.Clear()
+
+	tableBuffer.Clear()
+
+	return nil
+}
+
+func (l *LogClient) restoreTableFromPuller(ctx context.Context, tableID int64, puller *cdclog.EventPuller) error {
+	for {
+		item, err := puller.PullOneEvent(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if item == nil {
+			log.Info("[restoreFromPuller] nothing in puller, we should stop and flush")
+			err := l.applyKVChanges(ctx, tableID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			return nil
+		}
+		if !l.tsInRange(item.TS) {
+			log.Warn("[restoreFromPuller] ts not in given range, we should stop and flush",
+				zap.Uint64("start ts", l.startTS),
+				zap.Uint64("end ts", l.endTs),
+				zap.Uint64("item ts", item.TS),
+			)
+			err := l.applyKVChanges(ctx, tableID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			return nil
+		}
+
+		switch item.ItemType {
+		case cdclog.DDL:
+			name := l.meta.Names[tableID]
+			schema, table := parseQuoteName(name)
+			// ddl not influence on this schema/table
+			if schema != item.Schema || table != item.Table {
+				log.Info("[restoreFromPuller] meet unrelated ddl, and continue pulling",
+					zap.String("item table", item.Table),
+					zap.String("table", table),
+					zap.String("item schema", item.Schema),
+					zap.String("schema", schema),
+					zap.Int64("current table id", tableID),
+				)
+				continue
+			}
+			ddl := item.Meta.(*cdclog.MessageDDL)
+			log.Debug("[restoreFromPuller] exec ddl", zap.String("query", ddl.Query))
+
+			if l.isCreateDDL(ddl.Query) {
+				// we already put all create ddl at beginning of log restore
+				// so we should skip when meet create ddl
+				log.Info("[restoreFromPuller] meet create ddl, and continue pulling",
+					zap.String("item table", item.Table),
+					zap.String("table", table),
+					zap.String("item schema", item.Schema),
+					zap.String("schema", schema),
+					zap.Int64("current table id", tableID),
+				)
+				continue
+			}
+
+			// wait all previous kvs ingest finished
+			err = l.applyKVChanges(ctx, tableID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			err = l.restoreClient.db.se.Execute(ctx, ddl.Query)
```

Comment:
After executing ddl job, maybe table columns and column permutations should also be updated?